src/reporting/claim-reporting-fees.js

```diff
@@ -0,0 +1,165 @@
+ */
+
+ /**
+ * @typedef {Object} gasEstimateInfo
+ * @property {Array.<Object>} disputeCrowdsourcers Array of objects containing contract address/gas estimate pairs for all DisputeCrowdsourcers.
+ * @property {Array.<Object>} initialReporters Array of objects containing contract address/gas estimate pairs for all InitialReporters.
+ * @property {Array.<Object>} feeWindows Array of objects containing contract address/gas estimate pairs for all FeeWindows.
+ * @property {gasEstimateTotals} totals Object containing total gas estimates for each type of contract.
+ */
+
+ /**
+ * @typedef {Object} ClaimReportingFeesInfo
+ * @property {Array.<string>|null} redeemedFeeWindows Addresses of all successfully redeemed Fee Windows, as hexadecimal strings. Not set if `p.estimateGas` is true.
+ * @property {Array.<string>|null} redeemedDisputeCrowdsourcers Addresses of all successfully redeemed Dispute Crowdsourcers, as hexadecimal strings.  Not set if `p.estimateGas` is true.
+ * @property {Array.<string>|null} redeemedInitialReporters Addresses of all successfully redeemed Initial Reporters, as hexadecimal strings.  Not set if `p.estimateGas` is true.
+ * @property {gasEstimateInfo|null} gasEstimates Object containing a breakdown of gas estimates for all reporting fee redemption transactions. Not set if `p.estimateGas` is false.
+ * @property {Array.<RPCError|Error>} errors Array of errors returned when attempting to make transactions or get gas estimates, indexed by contract address.
+ */
+
+/**
+ * @param {Object} p Parameters object.
+ * @param {string} p.redeemer Ethereum address attempting to redeem reporting fees, as a hexadecimal string.
+ * @param {Array.<RedeemableContract>} p.redeemableContracts Array of objects containing contract address and contract type pairs.
+ * @param {boolean=} p.estimateGas Whether to return gas estimates for the transactions instead of actually making the transactions.
+ * @param {{signer: buffer|function, accountType: string}=} p.meta Authentication metadata for raw transactions.
+ * @param {function} callback Called after all transactions have been attempted.
+ * @return {ClaimReportingFeesInfo} Object containing information about which fees were successfully claimed or a breakdown of gas estimates.
+ */
+function claimReportingFees(p, callback) {
+  var redeemPayload = immutableDelete(p, ["redeemer", "redeemableContracts", "estimateGas"]);
+  var redeemedFeeWindows = [];
+  var redeemedDisputeCrowdsourcers = [];
+  var redeemedInitialReporters = [];
+  var gasEstimates = {
+    disputeCrowdsourcers: [],
+    initialReporters: [],
+    feeWindows: [],
+    totals: {
+      disputeCrowdsourcers: new BigNumber(0),
+      initialReporters: new BigNumber(0),
+      feeWindows: new BigNumber(0),
+      all: new BigNumber(0),
+    },
+  };
+  var errors = [];
+
+  async.eachLimit(p.redeemableContracts, PARALLEL_LIMIT, function (contract, nextContract) {
+    switch (contract.type) {
+      case 0:
```

Comment:
Do we have constants for these contract types?

Reply:
Sounds good. I'll add this.